hack/generator/pkg/config/configuration.go

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package config
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/Azure/k8s-infra/hack/generator/pkg/astmodel"
+)
+
+// Configuration is used to control which types get generated
+type Configuration struct {
+	// Base URL for the JSON schema to generate
+	SchemaURL string
+	// Filters used to control which types are included
+	ExportFilters []*ExportFilter
+	// Filters used to control which types are included in the type graph during JSON schema parsing
+	TypeFilters []*TypeFilter
+	// TypeTransformers used to remap types
+	TypeTransformers []*TypeTransformer
+}
+
+// ShouldExportResult is returned by ShouldExport to indicate whether the supplied type should be exported
+type ShouldExportResult string
+
+const (
+	// Export indicates the specified type should be exported to disk
+	Export ShouldExportResult = "export"
+	// Skip indicates the specified type should be skipped and not exported
+	Skip ShouldExportResult = "skip"
+)
+
+// ShouldPruneResult is returned by ShouldPrune to indicate whether the supplied type should be exported
+type ShouldPruneResult string
+
+const (
+	// Include indicates the specified type should be included in the type graph
+	Include ShouldPruneResult = "include"
+	// Prune indicates the type (and all types only referenced by it) should be pruned from the type graph
+	Prune ShouldPruneResult = "prune"
+)
+
+// NewConfiguration is a convenience factory for Configuration
+func NewConfiguration(filters ...*ExportFilter) *Configuration {
+	result := Configuration{
+		ExportFilters: filters,
+	}
+
+	return &result
+}
+
+// Initialize checks for common errors and initializes structures inside the configuration
+// which need additional setup after json deserialization
+func (config *Configuration) Initialize() error {
+	if config.SchemaURL == "" {
+		return errors.New("SchemaURL missing")
+	}
+
+	for _, transformer := range config.TypeTransformers {
+		err := transformer.Init()
+		if err != nil {
+			return err
+		}
+	}
```

Comment:
From a user experience perspective, I'm wondering whether we should collate errors together and return all of them at once; playing whack-a-mole fixing one error at a time can get pretty tedious.

Reply:
Hmm okay, so I did this but of course the actual functionality to do this easily isn't built into the language so I added a dependency which supports it.

I'm not 100% sure that's ok so will need to check around about what the process is (i.e. license of dependency, etc...)

Reply:
Poked around and confirmed that was okay